### ISSUE TITLE
AZURE DNN website multiple Instance issue

### ISSUE THREAD

**Issue by mailnaga** [davidjrh/dnn.rediscachingprovider#3]
I have a DNN website on AZURE with multiple portals. DNN version 7.3.1, I installed rediscaching provider and it worked fine until i had one instance. 

![image](https://cloud.githubusercontent.com/assets/8998913/9018991/4374ccd6-379a-11e5-804b-4c3dd4933aa1.png)

but when i increase the number of instances in AZURE i get 404 s on various pages, when i restart the website get 404 on different pages which shows its an web farm issue.

Please let know if i need to try anything else.

Thanks in advance
Naga

**Comment by davidjrh** [comment]
Hi, I've tested the scenario again and seems to be working fine. I used:
- DNN Platform 7.4.1 website with multiple portals
- Redis Caching Provider v1.0.3
- Started with 1 instance browsing the Host/Dashboard page (i.e. /Host/Dashboard/portalid/0), and check the "Server Name".
- Scaled up to 2 instances, and after a while, revisited the page to check the server name to see that was a different one. Note that you need to clear the ARRAffinityCookie from your browser or you will always hit the same instance. I did it with an extension in Chrome. 
- Repeated the same steps after scaling to more instances and saw that different servers were serving the pages with no 404 errors

Can you check the log4net files under "\Portals\_default\Logs" and see if there is any additional info there? Perhaps the provider is not setup correctly.

**Comment by mailnaga** [comment]
Thanks for checking this David.

I see the below in the event log 

![image](https://cloud.githubusercontent.com/assets/8998913/9142354/ff198932-3cf3-11e5-8e5f-041a70cab328.png)

below is the code that its pointing to, the object reference error is on the GetTabModules which inturn is going to DotNetNuke.Entities.Modules.ModuleController.

![image](https://cloud.githubusercontent.com/assets/8998913/9142531/1b90d678-3cf5-11e5-80f9-beea8fe954d5.png)

GetTabModules in DNN ModuleController does the below, which tries to get it from cache and this empty which means the cache is not massaged between the webfarm right? which does not happen with one instance.

![image](https://cloud.githubusercontent.com/assets/8998913/9142673/2701c160-3cf6-11e5-8a24-404f6f261ee6.png)

Looking forward to hear from  you.

Thanks for your responses.

**Comment by davidjrh** [comment]
Which URL provider are you using? If using the DNNFriendlyUrl, which urlFormat? Can you check if changing from urlFormat="humanfriendly" to urlFormat="advanced" makes any difference?

**Comment by mailnaga** [comment]
Thanks for you response. there is an Extension provider that we are using for the site. trying to setup an sandbox environment for you check the issue.

**Comment by davidjrh** [comment]
Can be that this Url extension is not correctly clearing the cache after adding/removing pages from the portal, so when the task is done in one of the servers, the others doesn't know anything of the new page until they are notified. While you setup the sandbox, can you test if anything of the following works?
- Manually clearing the cache by using the "Tools > Clear cache" menu
- Use the DNNFriendlyUrl provider

**Comment by mailnaga** [comment]
Manually clearing the cache will bring that page up but some other page will start returning 404
It is pointing to DNNFriendlyUrl provider in web config but there is Extension Url provider that we use for Product urls.

Can you give us your msdn or any Microsoft account email address so that we shall give access on AZURE.

**Comment by davidjrh** [comment]
Hi, I found the issue causing the out of sync cache. On parametrized ClearCache calls, the "type" and "data" parameters were not being published on the Redis channel, so were not effectively being cleared on the other webservers. 

![image](https://cloud.githubusercontent.com/assets/489976/12212046/b80e11c8-b65d-11e5-9f34-a63e74212d06.png)


I have added those parameters and now is working properly.I have created a new release 1.0.4 with the fix.
![image](https://cloud.githubusercontent.com/assets/489976/12212071/e9a9685e-b65d-11e5-87cf-57e9004453c7.png)
